Insert DefaultBrowser Cta and SearchWidget Cta as Dax dialog during Dax Journey

cmonfortep opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/488551667048375/1161278038072567/f
Tech Design URL: 
CC: 

**Description**:
In this experiment we are going to place our CTAs inside our Dax Journey. Our CTAs will match the style/tone of the existing Dax dialogs.

- Default Browser CTA: After the user visits a site with trackers, we prompt a Dax dialog to tell the user that we blocked all the trackers in that site. After showing that dialog we will ask the user to set us as Default browser.

- Search Widget CTA: When the user visits DuckDuckGo, a Dax dialog will appear informing our users that all their searches are private. After showing that dialog we will ask the user to add our search widget to protect their future searches.

**Concept test variant mj: Vanila concept test**

1. Hardcode into variantManager variant **"mj"**
1. Perform a fresh install
1. Run the app and ensure onboarding doesn't show any cta
1. Visit a web with trackers like cnn.com
1. Ensure dax dialog appears informing about trackers found
1. Click on "High five!"
1. Ensure no more Ctas are shown
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure no more Ctas are shown

**Concept test variant ml: Concept test with old ctas**

1. Hardcode into variantManager variant **"ml"**
1. Perform a fresh install
1. Run the app and ensure during the onboarding Default browser cta appears
1. Visit a web with trackers like cnn.com
1. Ensure dax dialog appears informing about trackers found
1. Click on "High five!"
1. Ensure no more Ctas are shown
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure no more Ctas are shown
1. Continue the Dax journey till the end (open new tab, visit a website, open a new tab)
1. Ensure Search widget cta appears in home screen

**Concept test variant mh: concept test with ctas as dax dialogs**:

> Dax Default Browser cta:

- Test-1

1. Hardcode into variantManager variant **"mh"**
2. Perform a fresh install
3. Run the app and ensure onboarding doesn't show any default browser cta
4. Visit a web with trackers like cnn.com
5. Ensure dax dialog appears informing about trackers found
6. Click on "High five!"
7. Ensure DefaultBrowser cta appears 
8. Ensure pixel "m_odc_s" is fired with "dbd" as param
9. Click on "Ok"
10. Ensure pixel "m_odc_ok" is fired with "dbd" as param
11. Ensure Dialog asking to set ddg as default browser appears

_Repeat each of the following scenarios with previous steps:_

a) 
12. Click on "just once"
13. Ensure dialog appears again
14. Ensure no pixel is fired
14. Click on "always"
15. Ensure dialog dissapears
16. Ensure pixel "m_db_s" is fired with "dbo=d" as param 

b) 
12. Click on "just once"
13. Ensure dialog appears again
14. Click on "always"
15. Ensure dialog dissapears
16. Ensure pixel "m_db_ns" is fired with "dbo=jom" as param 

c) 
12. press back
13. Ensure dialog is dismissed
14. Ensure pixel "m_db_ns" is fired with "dbo=dd" as param 

d) 
12. select another browser
13. Go back to DDG app
14. Ensure dialog is dismissed
14. Ensure pixel "m_db_ns" is fired with "dbo=e" as param 

- Test-2

1. Hardcode into variantManager variant **"mh"**
2. Perform a fresh install
3. Go to settings and set as another browser as default
4. Run the app and ensure onboarding doesn't show any default browser cta
5. Visit a web with trackers like cnn.com
6. Ensure dax dialog appears informing about trackers found
7. Click on "High five!"
8. Ensure DefaultBrowser cta appears 
9. Ensure pixel "m_odc_s" is fired with "dbs" as param


_Repeat each of the following scenarios with previous steps:_

a)
10. Click on "Go to settings"
11. Ensure pixel "m_odc_ok" is fired with "dbs" as param
12. Ensure user navigates to settings
13. Select ddg as default browser
14. Go back to the app
15. Ensure pixel "m_db_s" is fired with "fo=false, dbo=dxs" as param
16. Ensure cta is gone

b)
10. Click on "Maybe later"
11. Ensure pixel "m_odc_c" is fired with "cta=dbs" as param
12. Ensure cta is gone


> Dax Search widget cta:

- Test-1

1. Hardcode into variantManager variant **"mh"**
1. Perform a fresh install
1. Run the app and ensure onboarding doesn't show any default browser cta
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure SearchWidget cta appears 
1. Ensure pixel "m_odc_s" is fired with "wa" as param
1. Click on "Add"
1. Ensure pixel "m_odc_ok" is fired with "wa" as param
1. Ensure cta is gone
1. Ensure dialog appears asking the user to add the search widget

- Test-2

1. Hardcode into variantManager variant **"mh"**
1. Perform a fresh install
1. Run the app and ensure onboarding doesn't show any default browser cta
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure SearchWidget cta appears 
1. Ensure pixel "m_odc_s" is fired with "wa" as param
1. Click on "Maybe later"
1. Ensure pixel "m_odc_c" is fired with "wa" as param
1. Ensure cta is gone

- Test-3 (device running api < 23)

1. Hardcode into variantManager variant **"mh"**
1. Perform a fresh install
1. Run the app and ensure onboarding doesn't show any default browser cta
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure no cta appears

- Test-4 (device running api 23/24/26)

1. Hardcode into variantManager variant **"mh"**
1. Perform a fresh install
1. Run the app and ensure onboarding doesn't show any default browser cta
1. Visit duckduckgo.com
1. Ensure dax dialog appears
1. Click on "Phew!"
1. Ensure SearchWidget cta appears 
1. Ensure pixel "m_odc_s" is fired with "wm" as param
1. Click on "Add"
1. Ensure pixel "m_odc_ok" is fired with "wm" as param
1. Ensure cta is gone
1. Ensure dialog appears showing the user how to add the search widget

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

marcosholgado commented on app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:
```diff
@@ -1465,75 +1473,157 @@ class BrowserTabViewModelTest {
+    }
+
+    @Test
+    fun whenUserClickedDefaultBrowserCtaButtonWithDefaultBrowserThenLaunchDialogCommand() {
```
Something weird going on with this test. We are saying the dialog comment is launched but asserting the command issued is `OpenSettings` rather than `OpenDialog`.

marcosholgado commented on app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:
```diff
@@ -1465,75 +1473,157 @@ class BrowserTabViewModelTest {
+    }
+
+    @Test
+    fun whenUserFailedSettingDdgAsDefaultBrowserFromDialogOnceThenFirePixelAndUpdateInstallStore() {
```
This test doesn't do what its saying

marcosholgado commented on app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelNextCtaTest.kt:
```diff
@@ -0,0 +1,209 @@
+import org.junit.experimental.theories.Theory
+import org.junit.runner.RunWith
+
+@RunWith(Theories::class)
```
Nice! Never seen the theories runner being used anywhere!

cmonfortep replied:
yeah! I wrote a parameterized test with too many cases, `Theories` can help to reduce the test cases and to focus only on the expected result in different scenarios.

marcosholgado commented on app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt:
```diff
@@ -55,6 +56,8 @@ import org.junit.Assert.*
 import org.junit.Rule
 import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
```
These two imports are not being used

marcosholgado commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -539,9 +548,26 @@ class BrowserTabFragment : Fragment(), FindListener, CoroutineScope {
+        defaultBrowserNavigation.openDefaultBrowserDialog(this, url, DEFAULT_BROWSER_REQUEST_CODE_DIALOG)
+    }
+
+    private fun hidInstructionsCard() {
```
Typo: `hideInstructionsCard`

marcosholgado commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -539,9 +548,26 @@ class BrowserTabFragment : Fragment(), FindListener, CoroutineScope {
+    private fun openDefaultBrowserDialog(url: String) {
+        instructionsCard?.show()
+        defaultCard?.show()
+        defaultCard?.animate()?.alpha(1f)?.setDuration(100)?.start()
```
Consider using a const for `100` so it can be reused when hiding the card as well.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -898,29 +905,78 @@ class BrowserTabViewModel(
 
-    fun onUserClickCtaOkButton() {
-        val cta = ctaViewState.value?.cta ?: return
+    fun onUserClickCtaOkButton(cta: Cta) {
```
I found out some race conditions when using the current state from ViewModel instead of the CTA value from the original listener.

For example:
* We prompt a dialog
* User clicks in the ok button
* Dialog triggers `onUserClickCtaOkButton`
* New logic triggers a new dialog and CTA ViewState gets updated
* Dialog gets dismissed and triggers `onUserDismissedCta`
* The wrong dialog will be tracked (ViewState contains the new dialog rather than the previous one)

There are more scenarios, but this already shows that using this mutable ViewModel-ViewState could have non-expected behaviors. I think is more correct to rely on the actual value received from the listener.

cmonfortep commented:
Thanks for the review @marcosholgado 
1. I added the logic to detect when the user has selected another browser from the dialog following you advice. Thanks, good point there.

2. AND
3. Changed the pixelName values to make them shorter, and also removed the ones I added. Now I use the existing ones from the onboarding. I updated the description fixing the cases I wrongly described and adding a new one for the external browser.

4. Fixed dismissing the dialog when the user clicks before the animation is finished. 

5. The bug related to the instructions card I left it out of this scope and I will create a side task for that.
